Dismiss Suggestions on Space Keystroke

SiobhyB opened the pull request:
Partial fix for the issues outlined in https://github.com/WordPress/gutenberg/issues/24425

The iOS side to this issue can be found at https://github.com/wordpress-mobile/WordPress-iOS/pull/16601

## Description

If a user enters an `@` symbol on mobile, the suggestions UI automatically opens. There are times when a user doesn't want this UI but, instead, only wants to add a `@` symbol. This PR seeks to make it easier for users to exit the suggestions UI, by automatically closing it if they press a space key immediately after adding a `@`.

## Testing Steps

The testing steps match [the iOS side](https://github.com/wordpress-mobile/WordPress-iOS/pull/16601) of this issue:

* Verify that in the block editor, typing a space after a `@` character dismisses the mentions suggestions list
* Verify that in the block editor, typing a space after a `+` character dismisses the cross-posting suggestions list
* Verify that mentions still works, ensuring that the intended recipient gets notified
* Verify that cross-posting still works, ensuring that the intended recipient gets notified

In addition, this PR includes some [changes to the way Tracks events are called from `finishWithValue()`](https://github.com/wordpress-mobile/WordPress-Android/pull/14793/commits/9459b6cd841d4995c4b862da97261a320436b882). We should test that the event fires as expected, by following these steps: 

* Navigate to the **Live View** section of the Tracks tool. Enter the username of the account you're testing with and use search for the `wpandroid_suggestion_session_finished` event name. Note, it can take anywhere from five minutes to two hours for an event to be recorded.
* Verify that the `did_select_suggestion` is `true` in cases where you select the suggestion in your testing and `false` in cases where you exist the UI via the space keystroke.

P2 detailing the Tracks event, for reference: p9ugOq-1tR-p2

## Regression Notes
1. Potential unintended areas of impact

* Any new breaking or unusual behaviour with the existing suggestions UI.
* Incorrect values returning for the `did_select_suggestion` property in the Tracks event i.e. `true` when the user exits the UI without making a selection or `false` when they make a selection.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

I followed the testing steps listed above.

3. What automated tests I added (or what prevented me from doing so)

As this is a small change and isn't a core flow, a UI test doesn't feel warranted.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

SiobhyB committed:
Add conditional check for space keystroke

With this commit, a check has been added to the "onTextChanged" function. This function fires after a user begins entering text into the suggestions UI. The new conditional checks for whether a user has pressed the space key immediately after the UI appears.

SiobhyB committed:
Exit UI and update editor with space keystroke

With this commit, an empty space is passed to the "finishWithValue()" function when a user presses space key immediately when the suggestions UI opens.

"finishWithValue()" serves to exit the UI and update the editor with the updated value (in our case, an empty space key).

SiobhyB committed:
Update Tracks event to represent exit

When the suggestions UI is exited, a value of "false" should be sent back with the tracks event. See: p9ugOq-1tR-p2

With this commit, an if/else statement is added to "finishWithValue()" to send that value of "false" in cases where only a space has been passed to the function (indicating the user has simply existed the UI).

SiobhyB committed:
Merge branch 'develop' into gutenberg/24425-dismiss-suggestions-on-space-keystroke

SiobhyB committed:
Merge branch 'develop' into gutenberg/24425-dismiss-suggestions-on-space-keystroke

SiobhyB commented:
@mchowning, the iOS side to this issue has been updated in https://github.com/wordpress-mobile/WordPress-iOS/pull/16770 and now matches the approach to analytics being taken in this PR. Would you be able to review this when you have the chance? Thanks in advance!

mchowning commented on WordPress/src/main/java/org/wordpress/android/ui/suggestion/SuggestionActivity.kt:
```diff
@@ -201,7 +203,12 @@ class SuggestionActivity : LocaleAwareActivity() {
+            viewModel.trackExit(false)
+        } else {
+            viewModel.trackExit(true)
+        }
```
I would write this just slightly differently myself. One change worth considering is instead of the if statement, we can just define a `shouldTrackExit` variable:

```
val shouldTrackExit = value == ""
viewModel.trackExit(shouldTrackExit)
```

Or we could take that further and just inline the variable:

```
viewModel.trackExit(value == "")
```

I would actually make a slightly bigger change myself though. The thing that jumps out to me is that the "logic" around handling "spaces" is split into two places.

1. In the `TextWatcher` we check for `prefix` plus a single space, and if we find that we call `finishWIthValue` with an empty string.
2. In `finishWithValue` we check for the empty string and use that to determine what to pass to `trackExit`.

Slightly better imo is to instead keep the logic around the "space handling" all in the `TextWatcher`. The way that comes to mind for doing this is to instead pass a boolean for what should be passed to `trackExit` as an argument to `finishWithValue`. Kotlin lets us avoid having to pass that argument everywhere we call `finishWithExit` by making it default to true. That way we only need to explicitly include the argument when we want it to be false (which is only in the one call we're adding in this PR).

```
...

    } else if (s.startsWith("$prefix ")) {
        finishWithValue("", false)
    }

...

private fun finishWithValue(value: String?, withSuggestion: Boolean = true) {
    viewModel.trackExit(withSuggestion)
  ...
}

```

To be clear, what you have here is fine, so let me know if you disagree or think I might be overlooking something. This is certainly not a critical or blocking change. Like I mentioned before, the main benefit of this change imo is that it puts all the "logic" relating to the space handling (what value to finish with and whether we want to track this as an exit with a suggestion) in a single location, which I think will make it easier to understand for someone looking at this in the future.

SiobhyB replied:
@mchowning, thank you so much for taking the time to clearly outline your thinking in such a clear way. 

I can see how defining a boolean over using an if statement makes the code more readable, and will take that lesson with me. 

I also agree that your approach of keeping the logic in one place as much as possible is much easier to parse (again, another lesson that I'll aim to apply in future code!) 

I've implemented the changes in https://github.com/wordpress-mobile/WordPress-Android/pull/14793/commits/681a231fa53e424d8d6a69c960fa339453867342.

SiobhyB committed:
Refactor logic around "space handling"

This commit refactors the logic that handles how a space is handled in the suggestions UI, as well as how that space is tracked.

The reason for this refactor is that the logic was previously split up into two different places, making it potentially tricky for future devs to parse. With this commit, it's handled in the "TextWatcher", with a boolean being passed to the "finishWithValue" function.

mchowning reviewed (approved):
Looks good and works well! Great job @SiobhyB !

SiobhyB commented:
Thanks so much, Matt!
